compiler/sem/semstmts.nim

```diff
@@ -480,6 +480,90 @@ proc copyExcept(n: PNode, i: int): PNode {.inline.} =
   for j in 0..<n.len:
     if j != i: result.add(n[j])
 
+proc tryMacroPragma(c: PContext, pragmas: ptr PNode, i: int,
+                    operand: PNode, flags: TExprFlags): PNode =
+  ## If the `i`-th pragma in the `pragmas` list is a macro/template pragma,
+  ## applies it to the untyped `operand` AST and returns the transform (or an
+  ## error). `nil` is returned otherwise.
+  ##
+  ## `pragmas` needs to point to the ``nkPragma`` AST within the `operand`
+  ## AST. `flags` are the flags passed to the after-call expansion.
+  ##
+  ## Example:
+  ##
+  ## ..code-block:: nim
+  ##
+  ##   let l {.m(a, b), rest.} # the operand AST
+  ##   # is transformed into
+  ##   m(a, b, let l {.rest.})
+  ##
+  let n = pragmas[]
+  assert n.kind == nkPragma
+  let
+    prag         = n[i]
+    isCallSyntax = prag.kind in nkPragmaCallKinds and prag.len >= 1
+    key =
+      if isCallSyntax: prag[0]
+      else:            prag
+
+  # we only want to process macro pragmas (ast transforms)
+  # xxx: @saem: this dance to filter through them is another sign that pragmas
+  #      shouldn't be associated to symbols as the syntax hints
+  if whichPragma(prag) == wInvalid and key.kind in nkIdentKinds:
+    # a custom pragma as opposed to a built-in
+    let (ident, err) = considerQuotedIdent(c, key)
+    if err != nil:
+      # don't report the error yet, a macro/template pragma could
+      # still make the AST well-formed
```

Comment:
This would then be an ordering violation? where further a macro pragma(s), if it exists, should have come first?

The biggest drawback I can think of is a spelling mistake in one pragma might go unreported because a second pragma clobbered it, leading to unexpected semantics.

The benefit of reporting the error, is that we can assume all pragma prior to the current `i` to have been fully applied, simplifying the concept of progress.

Reply:
I'm not sure I understand the question with the ordering violation, but with the spelling mistake part I do somewhat agree.

Other invalid AST in pragma lists is already ignored during the early annotation processing, so my thinking was that making an exception for invalid accent quotations is unexpected/inconsistent.

Reply:
Your edit helped, thanks. Hm, yeah, I see what you mean. The perspective I made the change from is that macro pragmas take precedence over anything else, operating on the unprocessed definition AST.

Reply:
Sorry it wasn't clear from the get go, I was more tired than I realized and ended up with some word salad. Glad to hear the edit helped clarify things.